### PR TITLE
Point to immutable arg/fields when trying to use as &mut

### LINKED ISSUE #38147: Improve E0389 "cannot borrow data mutably in a `&` reference" error message
Error in next case could be more helpful
```
struct Foo<'a> {
    s: &'a mut String
}

impl<'a> Foo<'a> {
    fn f(&self) {
        self.s.push('x');
    }
}

fn main() {}
```

error:
```
error[E0389]: cannot borrow data mutably in a `&` reference
 --> 2.rs:7:9
  |
7 |         self.s.push('x');
  |         ^^^^^^ assignment into an immutable reference
```

cc @jonathandturner

### PR DESCRIPTION
**Pull request by estebank**
Present the following output when trying to access an immutable borrow's
field as mutable:

```
error[E0389]: cannot borrow data mutably in a `&` reference
  --> $DIR/issue-38147-1.rs:27:9
   |
26 | fn f(&self) {
   |      -----  use `&mut self` here to make mutable
27 |     f.s.push('x');
   |     ^^^ assignment into an immutable reference
```

And the following when trying to access an immutable struct field as mutable:

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
   |
12 |     s: &'a String
   |     ------------- use `&'a mut String` here to make mutable
...|
16 |     fn f(&self) {
   |          -----  use `&mut self` here to make mutable
17 |         self.s.push('x');
   |         ^^^^^^ cannot borrow as mutable
```

Fixes #38147.

### CODE REVIEW

**Comment by estebank** [comment]
CC @jonathandturner @GuillaumeGomez @nikomatsakis

File: src/librustc_borrowck/borrowck/gather_loans/mod.rs [hunk 1]
```diff
@@ -177,7 +177,7 @@ fn check_aliasability<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>,
     debug!("check_aliasability aliasability={:?} req_kind={:?}",
            aliasability, req_kind);
 
-    match (aliasability, req_kind) {
+    match (aliasability.clone(), req_kind.clone()) {
```
Comment (GuillaumeGomez) [inline src/librustc_borrowck/borrowck/gather_loans/mod.rs]: Why suddenly cloning them?

**Comment by GuillaumeGomez** [comment]
This is great! 👍

**Comment by nrc** [comment]
r? @nikomatsakis (I'm not familiar enough with these bits of the compiler any more).

I would prefer less assertive language in the error messages - rather than "X should be Y", I prefer, "if you want X, then you can change X to Y" or something similar. Also, could these be `suggestions` rather than labels? One day I would like to be able to auto-apply suggestions like this and it makes it easier if they are declared as such.

File: src/librustc_borrowck/borrowck/gather_loans/mod.rs (hunk 1 above)
Comment (estebank) [inline src/librustc_borrowck/borrowck/gather_loans/mod.rs]: Leftover from previous exploration.

**Comment by estebank** [comment]
Introduced an error in [two tests](https://travis-ci.org/rust-lang/rust/jobs/192896881#L3110) (removed two hints). Will try to fix it tonight.

**Comment by estebank** [comment]
> I would prefer less assertive language in the error messages - rather than "X should be Y", I prefer, "if you want X, then you can change X to Y" or something similar.

@nrc, I'm ok with doing so. For now, I copied the message already being used for the same kind of suggestion:

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
   |
12 |     s: &'a String
   |     ------------- this field should be `&mut`
...|
16 |     fn f(&self) {
   |          -----  use `&mut self` here to make mutable
17 |         self.s.push('x');
   |         ^^^^^^
```

> Also, could these be suggestions rather than labels? One day I would like to be able to auto-apply suggestions like this and it makes it easier if they are declared as such.

The only things that makes me uneasy about using suggestions are that the actual output turns out to be slightly longer, and because of the out of order and lack of line number, less comprehensible:

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
   |
12 |     s: &'a String
   |     ------------- this field should be `&mut`
...
17 |         self.s.push('x');
   |         ^^^^^^
   |
help: try making this a mutable borrow
   |     fn f(&mut self) {
```

I just filed #39152 so it can be improved going forward, but as it is now I really prefer the `span_label` presentation.

**Comment by jonathandturner** [comment]
I'm definitely liking this revision.  One question, is this part missing something?

```
17 |         self.s.push('x');
   |         ^^^^^^
```

Seems like it needs a label there

**Comment by estebank** [comment]
@jonathandturner the [current output](https://is.gd/bfe41L) doesn't set a label. To add one, I could definitely use some help with the wording. Something along the lines of:

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
17 |         self.s.push('x');
   |         ^^^^^^ cannot borrow as mutable
```

or

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
17 |         self.s.push('x');
   |         ^^^^^^ cannot borrow immutable borrowed content as mutable
```

seems reasonable.

**Comment by jonathandturner** [comment]
@estebank choice #1 sounds good :)

**Comment by estebank** [comment]
@jonathandturner done.

**Comment by nikomatsakis** [comment]
👍 to shorter wording on the "primary label". In fact, though I think this is a separate thing, I have long wanted to aggressively shorten the borrowck messages, e.g. removing the phrase "borrowed as mutable" and just say "write" or "mutate", and definitely not talking about distracting but complex things like "immutable borrowed content".

Also 👍 to this general pull request. =)

**Comment by estebank** [comment]
@nikomatsakis is there anything else that should be part of this PR?

File: src/librustc_borrowck/borrowck/mod.rs [hunk 2]
```diff
@@ -913,6 +927,43 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
         err.emit();
     }
 
+    fn immutable_argument_should_be_mut(&self, nid: ast::NodeId, db: &mut DiagnosticBuilder) {
+        let parent = self.tcx.map.get_parent_node(nid);
+
+        if let Some(fn_like) = FnLikeNode::from_node(self.tcx.map.get(parent)) {
```
Comment (nikomatsakis) [inline src/librustc_borrowck/borrowck/mod.rs]: So...this fn is pretty hard to read. =) Maybe we can use some temporaries or something? I just find it hard to follow, but maybe it's ok, if there was at least a comment. Basically when I read it I just see "walk a bunch of data structures to pull out data", but my eyes slide over the details...

File: src/test/ui/did_you_mean/issue-38147-2.stderr [hunk 3]
```diff
@@ -0,0 +1,11 @@
+error: cannot borrow immutable borrowed content `*self.s` as mutable
+  --> $DIR/issue-38147-2.rs:17:9
+   |
+12 |     s: &'a String
+   |     ------------- this field should be `&mut`
```
Comment (nikomatsakis) [inline src/test/ui/did_you_mean/issue-38147-2.stderr]: nice :)

File: src/test/ui/did_you_mean/issue-38147-3.stderr [hunk 4]
```diff
@@ -0,0 +1,13 @@
+error: cannot borrow immutable borrowed content `*self.s` as mutable
+  --> $DIR/issue-38147-3.rs:17:9
+   |
+12 |     s: &'a String
+   |     ------------- this field should be `&mut`
+...
+16 |     fn f(&self) {
+   |          ----- use `&mut self` here to make mutable
```
Comment (nikomatsakis) [inline src/test/ui/did_you_mean/issue-38147-3.stderr]: maybe we can adjust the wording to make it clearer that BOTH parts need to be `&mut self`? That's a bit tricky, particularly since we don't control the relative order of things here

File: src/test/ui/span/borrowck-call-is-borrow-issue-12224.stderr [hunk 5]
```diff
@@ -17,13 +17,18 @@ error: cannot borrow immutable borrowed content `*f` as mutable
...
    |                -- use `&mut F` here to make mutable
 36 |     (*f)(); //~ ERROR: cannot borrow immutable borrowed content `*f` as mutable
-   |     ^^^^
+   |     ^^^^ cannot borrow as mutable
 
 error: cannot borrow immutable `Box` content `*f.f` as mutable
   --> $DIR/borrowck-call-is-borrow-issue-12224.rs:44:5
    |
+18 |     f: Box<FnMut() + 'a>
+   |     -------------------- this field should be `&mut`
```
Comment (nikomatsakis) [inline src/test/ui/span/borrowck-call-is-borrow-issue-12224.stderr]: Hmm, this seems wrong, no?

Comment (nikomatsakis) [inline src/test/ui/span/borrowck-call-is-borrow-issue-12224.stderr]: Certainly there is nothing that would easily be converted to `&mut` here

File: src/librustc_borrowck/borrowck/mod.rs [hunk 6]
```diff
@@ -830,13 +834,20 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
                          its contents can be safely reborrowed",
                         descr)
             }
+        };
+
+        let mut db = self.struct_span_err(span, msg);
+        if let Some(span) = field_span {
+            db.span_label(span, &"this field should be `&mut`");
```
Comment (nikomatsakis) [inline src/librustc_borrowck/borrowck/mod.rs]: seems like we have to check here that the type of the field is some sort of reference, right? (hence the "over fire" case in the `ui` tests below)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks like it's close but not quite there. We should make some tests targeting this specific case (the existing tests seem to have caught this overenthusiastic error message "by accident", no?)

File: src/test/ui/span/borrowck-call-is-borrow-issue-12224.stderr (hunk 5 above)
Comment (estebank) [inline src/test/ui/span/borrowck-call-is-borrow-issue-12224.stderr]: What should be the appropriate message for this case? Nothing if the type is not a ref?

File: src/librustc_borrowck/borrowck/mod.rs (hunk 2 above)
Comment (estebank) [inline src/librustc_borrowck/borrowck/mod.rs]: This is moved code from bellow so that I could use it twice without copying code around :)

I will see how I can make it easier to follow.

File: src/test/ui/did_you_mean/issue-38147-3.stderr (hunk 4 above)
Comment (estebank) [inline src/test/ui/did_you_mean/issue-38147-3.stderr]: I could look at the span's line number to identify order and change this to something like

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> $DIR/issue-38147-3.rs:17:9
   |
12 |     s: &'a String
   |     ------------- this field should be `&mut`...
...
16 |     fn f(&self) {
   |          ----- ...and use `&mut self` here to make mutable
17 |         self.s.push('x');
   |         ^^^^^^ cannot borrow as mutable
```

But I don't think it's unreasonable to show both labels, and if the developer only changes one of them, to still get the same error with the remaining label.

**Review by estebank — COMMENTED** [review]
> We should make some tests targeting this specific case

issue-38147-1, issue-38147-2, issue-38147-3 & issue-38147-4 are new tests targeting this case.

**Comment by nikomatsakis** [comment]
@estebank 

> What should be the appropriate message for this case? Nothing if the type is not a ref?

Yeah, no change is needed in this case.

**Comment by nikomatsakis** [comment]
@estebank 

Travis is failing because a UI test needs to be updated:

```
error: cannot borrow immutable borrowed content `*self.s` as mutable
  --> /checkout/src/test/ui/did_you_mean/issue-38147-3.rs:17:9
   |
12 |     s: &'a String
   |     ------------- use `&'a mut String` here to make mutable
...
16 |     fn f(&self) {
   |          ----- use `&mut self` here to make mutable
17 |         self.s.push('x');
   |         ^^^^^^ cannot borrow as mutable

error: aborting due to previous error
```

**Comment by nikomatsakis** [comment]
but the new output looks correct =)

**Comment by estebank** [comment]
@nikomatsakis tests are fixed. Can you check wether the method [`immutable_argument_should_be_mut`](https://github.com/rust-lang/rust/pull/39139/files#diff-8c516b1674199747f85038166dd95527R968) is now more readable (along with the new method [`suggest_mut_for_immutable `](https://github.com/rust-lang/rust/pull/39139/files#diff-8c516b1674199747f85038166dd95527R940))?

Also, please verify [that using `bug!` for a case that *shouldn't ever* happen](https://github.com/rust-lang/rust/pull/39139/files#diff-8c516b1674199747f85038166dd95527R962) is correct. If it is not, it can be changed to return a generic comment `use '&mut' to make mutable`.

**Comment by nikomatsakis** [comment]
@estebank 

> Can you check wether the method immutable_argument_should_be_mut is now more readable (along with the new method suggest_mut_for_immutable)?

Yes; it's a wonder what some comments will do! :) Thanks.

> Also, please verify that using bug! for a case that shouldn't ever happen is correct. If it is not, it can be changed to return a generic comment use '&mut' to make mutable.

Yes that is correct.

**Review by nikomatsakis — APPROVED** [review]

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfaulted https://github.com/rust-lang/rust/issues/38878

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfaulted https://github.com/rust-lang/rust/issues/38878

**Comment by alexcrichton** [comment]
@bors: retry

* presumed bug in travis

**Comment by arielb1** [comment]
This causes the ICE #39544.
